### ISSUE TITLE
Setting GT (genotype) and AD (allele depths)

### ISSUE THREAD

**Issue by nh13** [varlociraptor/varlociraptor#153]
I am performing tumor-normal calling and I'd like the output VCF to contain both a genotype (FORMAT/GT) and per-allele depth (FORMAT/AD).  Many downstream tools will want these (I can't change the world in one day, but I can get folks to use this tool!).  I'd be glad to hear folks' thoughts on the below.

#### Allele Depths

The `FORMAT/OBS` field looks very informative, providing the summary of observations per _alternate allele_.  To get the per-alternate-allele depths, I could pick a cutoff (strong?) and assign the read counts to the particular allele.  I can then get the _reference_ allele depth by summing up the per-alternate-allele depths and subtracting that sum from the `FORMAT/DP` field.   This seems a bit risky, as I'd rather ALSO have the same summary CBTSOP for the reference allele.

Is there a better way (or perhaps new feature), since skimming the code it looks like the observations (both from evidence or pileups) are available?  With some guidance, I could see a contribution being made if the tool itself could output allele depths (or predicted allele depths).

#### Genotypes

For the `normal` sample's genotype, I think I can create the following `events` in the grammar:
```
  normal_hom_ref: "normal:0.0"
  normal_het: "normal:0.5"
  normal_hom_alt: "normal:1.0"
```
I can then pick the maximum posterior even for the `normal`.

For the `tumor` sample's genotype, I could do something like:
```
  somatic_tumor_high: "normal:0.0 & tumor:[0.1,1.0]"
  somatic_tumor_low: "normal:0.0 & tumor:]0.0,0.1["
  tumor_hom_ref: "tumor:0.0"
  tumor_het: "tumor:0.5"
  tumor_hom_alt: "tumor:1.0"
```
and pick from the scenarios above.   In case of a tie, I'd prefer the `somatic_tumor*` cases.

**Comment by dlaehnemann** [comment]
**Allele Depths**

Another idea would be to use the allele-specific `AF` and multiply this with the `DP`:
https://github.com/varlociraptor/varlociraptor/blob/0bb8f7b216127332ea7850db397496864eef1149/src/calling/variants/calling.rs#L120-L121

This should give you good estimates of the `AD` of alternative alleles, but I'm not sure if simply subtracting their sum from DP will give you a usable reference `AD`. Maybe this would overestimate it, but @johanneskoester can surely give a definitive answer, here.

**Genotypes**

For the events, you have to be careful that their definitions don't overlap. If they do, certain areas of the event space will be included multiple times in the marginal likelihood, distorting your posterior probabilities. And also bear in mind, that a an `ABSENT` event (`absent: "normal:0.0 & tumor:0.0"`) is always included by default, so this frequency combination (all samples at `0.0`) should not be included in any of the events you define.

For your setup above (and assuming that the normal `universe: "0.0 | 0.5 | 1.0"` and the tumor `universe: "[0.0,1.0]"`) , I would suggest the events to be defined as:
```
somatic_het_low: "normal:0.0 & tumor:]0.0,0.1["
somatic_het_high: "normal:0.0 & tumor:[0.1,0.5]"
somatic_hom_alt: "normal:0.0 & tumor:]0.5,1.0]"
somatic_het_loss: "normal:0.5 & tumor:0.0"
somatic_hom_alt_loss: "normal:1.0 & tumor:0.0"
germline_het_tumor_het: "normal:0.5 & tumor:]0.0,0.5]"
germline_het_tumor_hom_alt: "normal:0.5 & tumor:]0.5,1.0]"
germline_hom_alt_tumor_het: "normal:1.0 & tumor:]0.0,0.5]"
germline_hom_alt_tumor_hom_alt: "normal:1.0 & tumor:]0.5,1.0]"
```
You can then create summary events for the different genotypes in the different samples, summarising all events for the respective frequencies, so `0.0 | 0.5 | 1.0` for normal and `0.0, ]0.0,0.5], ]0.5,1.0]` for tumor:
```
prob_germline_hom_ref = PROB_ABSENT + PROB_SOMATIC_HET_LOW + PROB_SOMATIC_HET_HIGH + PROB_SOMATIC_HOM_ALT
prob_germline_het = PROB_GERMLINE_HET_TUMOR_HET + PROB_GERMLINE_HET_TUMOR_HOM_ALT + SOMATIC_HET_LOSS
prob_germline_hom_alt = PROB_GERMLINE_HOM_ALT_TUMOR_HET + PROB_GERMLINE_HOM_ALT_TUMOR_HOM_ALT + SOMATIC_HOM_ALT_LOSS

tumor_hom_ref = PROB_ABSENT + SOMATIC_HET_LOSS + SOMATIC_HOM_ALT_LOSS
tumor_het = PROB_SOMATIC_HET_LOW + PROB_SOMATIC_HET_HIGH + PROB_GERMLINE_HET_TUMOR_HET + PROB_GERMLINE_HOM_ALT_TUMOR_HET
tumor_hom_alt = PROB_SOMATIC_HOM_ALT + PROB_GERMLINE_HET_TUMOR_HOM_ALT + PROB_GERMLINE_HOM_ALT_TUMOR_HOM_ALT
```
In each of the sets (normal and tumor), each of the events (plus the ABSENT event) have to appear exactly once. Then, you can take the maximum posterior sum in each set as your genotype. Does this make sense, in principle?

**Comment by nh13** [comment]
##### Allele Depths

For the estimated per-allele depths, that could work, but it would be better to get a definitive (integer) count per allele as a lot of folks have "variant allele must be supported by at least N reads" type of requirements already built-in to their tests (clinical diagnostics), that would need to be supported (of course we can push for change). 

I'd be grateful for any ideas on how to assign the piled up (and presumably re-aligned reads) to each allele so that the sum of the allele depths equal the total depth.  @johanneskoester any thoughts?

##### Genotypes

@dlaehnemann not only is this exactly what I wanted for my tumor-normal calling, but now you've got me thinking about using this tool for a lot of other use-cases where I can define non-overlapping events with this grammar.

Quick question: I'd need to use the `varlociraptor call variants generic` to define those events as I don't see a way to feed them into `varlociraptor call variants tumor-normal`.  Is there a default `scenario` YAML I can copy that mimics the default settings for the built-in `tumor-normal` scenario?

**Comment by nh13** [comment]
@dlaehnemann also, from `varlociraptor` I'll get two records (lines) when I feed in a site with REF=A and ALTS=C,G: one with A/C and one with A/G.  How do I evaluate the possibility of A/G?

**Comment by dlaehnemann** [comment]
**Allele Depths**

If you want discrete integer counts, your idea of going for using a cutoff on the `OBS` makes the most sense. This classifies each single read, so you get discrete decisions. You could either take all reads that have `V/v`ery strong evidence and `S/s`trong evidence for the alternative allele, or further include those with `P/p`ositive evidence. But I'll also point @johanneskoester to this issue to maybe provide further insight and ideas.

**Genotypes**

Yes, this whole idea of an event space is great, right? So I'm glad that my write-down helped in illustrating it.

As for usable YAML files, @johanneskoester created a collection of them for the docs:
https://varlociraptor.github.io/varlociraptor-scenarios/landing/

But I guess they could be linked to more prominently... :D

And regarding the two records: There should be `OBS` entries and `PROBS` in each row, right? varlociraptor splits up multiallelic records during preprocessing and then evaluates each possible alternative allele against the reference allele separately. But unless you genuinely have a multiallelic site (and not just some sequencing errors or other artifacts), this should easily be filtered down to one (or none) of the alternative alleles with a simple false discovery rate control or by filtering based on the posterior odds: https://varlociraptor.github.io/docs/filtering/

And regarding pushing for change: This latter form of filtering is exactly what we are pushing for to use instead of arbitrary read count (or similar) filters. It's obviously never easy to get people to use a probabilistic approach that they might not fully comprehend over a very easily understandable threshold, but hopefully the idea of a false discovery rate is well-known and clear enough to get people to move to such a filtering...

Also, if you are looking for how we are currently using varlociraptor, our current best practices are developed in this snakemake workflow:
https://github.com/snakemake-workflows/dna-seq-varlociraptor

Feel free to use it directly or just nick ideas from it.

**Comment by nh13** [comment]
Oh, and my last comment becomes even more relevant when I have spanning deletions.  I have seen validated clinically relevant variants where we have a deletion on one haplotype that spans an SNV on the other haplotype (phased):
```
chr1 5 AAAAAA A ... 0|1
chr1 7 A C,* ... 2|1
```
The above shows that on one haplotype we have a `A>C` at `chr1:7`, and a deletion of 5bp starting at `chr1:5` on the other.

**Comment by dlaehnemann** [comment]
OK, I definitely see where these two cases are relevant, interesting and problematic -- and how they are challenging in different ways with the current parsing of records. I have created the separate issue #157 to keep track of this, then we can keep this issue here focused on the genotype and allele depth discussions. But please do add your thoughts over there, as well!

And yes, you understood correctly, that the relevant genotype `C/G` in your example above is not even evaluated with the current setup.

**Comment by johanneskoester** [comment]
@nh13 @dlaehnemann indeed, for reporting proper genotypes of multiallelic sites, there is some work to do. The way I planned it, would not require to modify the grammar though, it would still be able to just reason over VAFs. See my answer here: https://github.com/varlociraptor/varlociraptor/issues/157#issuecomment-831836859.

**Comment by johanneskoester** [comment]
Just a note here since I just discovered this in the discussion above. The whole point of Varlociraptor is to get rid of read count based filters. There should be no other filtering needed than what we do based on the event posteriors. So, supporting AD can be an addon for QC, but should definitely not be needed for filtering.
